{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SOView import *\n",
    "import pandas as pd\n",
    "import datetime\n",
    "import palettable\n",
    "import squarify\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_table_name = 'sodb_20220921.xlsx'\n",
    "data = pd.read_excel(data_table_name,sheet_name=0)\n",
    "dataset = pd.read_excel(data_table_name,sheet_name=1)\n",
    "biotech = pd.read_excel(data_table_name,sheet_name=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name_Short</th>\n",
       "      <th>Name_Long</th>\n",
       "      <th>Biotech_ID</th>\n",
       "      <th>doi</th>\n",
       "      <th>Date</th>\n",
       "      <th>Country</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Author</th>\n",
       "      <th>Access</th>\n",
       "      <th>Zip_Path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chen2021dissecting</td>\n",
       "      <td>Dissecting mammalian spermatogenesis using spa...</td>\n",
       "      <td>25</td>\n",
       "      <td>https://doi.org/10.1016/j.celrep.2021.109915</td>\n",
       "      <td>2021</td>\n",
       "      <td>USA</td>\n",
       "      <td>Cell Reports</td>\n",
       "      <td>Chen, H.\\nMurray, E.\\nSinha, A.\\nLaumas, A.\\nL...</td>\n",
       "      <td>https://www.dropbox.com/s/ygzpj0d0oh67br0/Test...</td>\n",
       "      <td>demo_data/zip/chen2021dissecting.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>stickels2020highly</td>\n",
       "      <td>Highly sensitive spatial transcriptomics at ne...</td>\n",
       "      <td>25</td>\n",
       "      <td>10.1038/s41587-020-0739-1</td>\n",
       "      <td>2020</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Biotechnology</td>\n",
       "      <td>Stickels, Robert R.\\nMurray, Evan\\nKumar, Pawa...</td>\n",
       "      <td>https://singlecell.broadinstitute.org/single_c...</td>\n",
       "      <td>demo_data/zip/maynard2021transcriptome.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>rodriques2019slide</td>\n",
       "      <td>Slide-seq: A scalable technology for measuring...</td>\n",
       "      <td>1</td>\n",
       "      <td>10.1126/science.aaw1219</td>\n",
       "      <td>2019</td>\n",
       "      <td>USA</td>\n",
       "      <td>Science</td>\n",
       "      <td>Rodriques, S. G.\\nStickels, R. R.\\nGoeva, A.\\n...</td>\n",
       "      <td>https://portals.broadinstitute.org/single_ cel...</td>\n",
       "      <td>demo_data/zip/hunter2021spatially.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>parigi2022the</td>\n",
       "      <td>The spatial transcriptomic landscape of the he...</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1038/s41467-022-28497-0</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Communications</td>\n",
       "      <td>Parigi, Sara M.\\nLarsson, Ludvig\\nDas, Srustid...</td>\n",
       "      <td>GSE190595</td>\n",
       "      <td>demo_data/zip/keren2018structured.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>guilliams2022spatial</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/j.cell.2021.12.018</td>\n",
       "      <td>2022</td>\n",
       "      <td>Belgium</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Guilliams, M.\\nBonnardel, J.\\nHaest, B.\\nVande...</td>\n",
       "      <td>GSE192742</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>84</td>\n",
       "      <td>117</td>\n",
       "      <td>Marshall2022High_mouse</td>\n",
       "      <td>High-resolution Slide-seqV2 spatial transcript...</td>\n",
       "      <td>25</td>\n",
       "      <td>https://doi.org/10.1016/j.isci.2022.104097</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>iScience</td>\n",
       "      <td>Marshall, Jamie L\\nNoel, Teia\\nWang, Qingbo S\\...</td>\n",
       "      <td>cellxgene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>85</td>\n",
       "      <td>118</td>\n",
       "      <td>Barkley2022Cancer</td>\n",
       "      <td>Cancer cell states recur across tumor types an...</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1038/s41588-022-01141-9</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Genetics</td>\n",
       "      <td>Barkley, D.\\nMoncada, R.\\nPour, M.\\nLiberman, ...</td>\n",
       "      <td>GSE203612</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>86</td>\n",
       "      <td>119</td>\n",
       "      <td>Biermann2022Dissecting</td>\n",
       "      <td>Dissecting the treatment-naive ecosystem of hu...</td>\n",
       "      <td>25</td>\n",
       "      <td>10.1016/j.cell.2022.06.007</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Biermann, J.\\nMelms, J. C.\\nAmin, A. D.\\nWang,...</td>\n",
       "      <td>GSE185386</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>87</td>\n",
       "      <td>120</td>\n",
       "      <td>Wang2018Three_1k</td>\n",
       "      <td>Three-dimensional intact-tissue sequencing of ...</td>\n",
       "      <td>9</td>\n",
       "      <td>10.1126/science.aat5691</td>\n",
       "      <td>2018</td>\n",
       "      <td>USA</td>\n",
       "      <td>Science</td>\n",
       "      <td>Wang, X.\\nAllen, W. E.\\nWright, M. A.\\nSylwest...</td>\n",
       "      <td>http://starmapresources.com/data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>88</td>\n",
       "      <td>121</td>\n",
       "      <td>Beechem2022High</td>\n",
       "      <td>High-plex Multiomic Analysis in FFPE at Subcel...</td>\n",
       "      <td>24</td>\n",
       "      <td>https://doi.org/10.1101/2021.11.03.467020</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>bioRxiv</td>\n",
       "      <td>Beechem, Joseph M</td>\n",
       "      <td>https://nanostring.com/products/cosmx-spatial-...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0   ID              Name_Short  \\\n",
       "0            0    0      chen2021dissecting   \n",
       "1            1    1      stickels2020highly   \n",
       "2            2    2      rodriques2019slide   \n",
       "3            3    3           parigi2022the   \n",
       "4            4    4    guilliams2022spatial   \n",
       "..         ...  ...                     ...   \n",
       "84          84  117  Marshall2022High_mouse   \n",
       "85          85  118       Barkley2022Cancer   \n",
       "86          86  119  Biermann2022Dissecting   \n",
       "87          87  120        Wang2018Three_1k   \n",
       "88          88  121         Beechem2022High   \n",
       "\n",
       "                                            Name_Long  Biotech_ID  \\\n",
       "0   Dissecting mammalian spermatogenesis using spa...          25   \n",
       "1   Highly sensitive spatial transcriptomics at ne...          25   \n",
       "2   Slide-seq: A scalable technology for measuring...           1   \n",
       "3   The spatial transcriptomic landscape of the he...           0   \n",
       "4   Spatial proteogenomics reveals distinct and ev...           0   \n",
       "..                                                ...         ...   \n",
       "84  High-resolution Slide-seqV2 spatial transcript...          25   \n",
       "85  Cancer cell states recur across tumor types an...           0   \n",
       "86  Dissecting the treatment-naive ecosystem of hu...          25   \n",
       "87  Three-dimensional intact-tissue sequencing of ...           9   \n",
       "88  High-plex Multiomic Analysis in FFPE at Subcel...          24   \n",
       "\n",
       "                                             doi  Date  Country  \\\n",
       "0   https://doi.org/10.1016/j.celrep.2021.109915  2021      USA   \n",
       "1                      10.1038/s41587-020-0739-1  2020      USA   \n",
       "2                        10.1126/science.aaw1219  2019      USA   \n",
       "3                     10.1038/s41467-022-28497-0  2022      USA   \n",
       "4                     10.1016/j.cell.2021.12.018  2022  Belgium   \n",
       "..                                           ...   ...      ...   \n",
       "84    https://doi.org/10.1016/j.isci.2022.104097  2022      USA   \n",
       "85                    10.1038/s41588-022-01141-9  2022      USA   \n",
       "86                    10.1016/j.cell.2022.06.007  2022      USA   \n",
       "87                       10.1126/science.aat5691  2018      USA   \n",
       "88     https://doi.org/10.1101/2021.11.03.467020  2022      USA   \n",
       "\n",
       "                  Journal                                             Author  \\\n",
       "0            Cell Reports  Chen, H.\\nMurray, E.\\nSinha, A.\\nLaumas, A.\\nL...   \n",
       "1    Nature Biotechnology  Stickels, Robert R.\\nMurray, Evan\\nKumar, Pawa...   \n",
       "2                 Science  Rodriques, S. G.\\nStickels, R. R.\\nGoeva, A.\\n...   \n",
       "3   Nature Communications  Parigi, Sara M.\\nLarsson, Ludvig\\nDas, Srustid...   \n",
       "4                    Cell  Guilliams, M.\\nBonnardel, J.\\nHaest, B.\\nVande...   \n",
       "..                    ...                                                ...   \n",
       "84               iScience  Marshall, Jamie L\\nNoel, Teia\\nWang, Qingbo S\\...   \n",
       "85        Nature Genetics  Barkley, D.\\nMoncada, R.\\nPour, M.\\nLiberman, ...   \n",
       "86                   Cell  Biermann, J.\\nMelms, J. C.\\nAmin, A. D.\\nWang,...   \n",
       "87                Science  Wang, X.\\nAllen, W. E.\\nWright, M. A.\\nSylwest...   \n",
       "88                bioRxiv                                  Beechem, Joseph M   \n",
       "\n",
       "                                               Access  \\\n",
       "0   https://www.dropbox.com/s/ygzpj0d0oh67br0/Test...   \n",
       "1   https://singlecell.broadinstitute.org/single_c...   \n",
       "2   https://portals.broadinstitute.org/single_ cel...   \n",
       "3                                           GSE190595   \n",
       "4                                           GSE192742   \n",
       "..                                                ...   \n",
       "84                                          cellxgene   \n",
       "85                                          GSE203612   \n",
       "86                                          GSE185386   \n",
       "87                   http://starmapresources.com/data   \n",
       "88  https://nanostring.com/products/cosmx-spatial-...   \n",
       "\n",
       "                                      Zip_Path  \n",
       "0         demo_data/zip/chen2021dissecting.zip  \n",
       "1   demo_data/zip/maynard2021transcriptome.zip  \n",
       "2        demo_data/zip/hunter2021spatially.zip  \n",
       "3        demo_data/zip/keren2018structured.zip  \n",
       "4                                          NaN  \n",
       "..                                         ...  \n",
       "84                                         NaN  \n",
       "85                                         NaN  \n",
       "86                                         NaN  \n",
       "87                                         NaN  \n",
       "88                                         NaN  \n",
       "\n",
       "[89 rows x 12 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 20,\n",
       " 22,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(biotech[['ID']].values[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "biotech_type_dict = dict(zip(biotech[['ID']].values[:,0],biotech[['Type']].values[:,0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'Spatial Transcriptomics',\n",
       " 1: 'Spatial Transcriptomics',\n",
       " 2: 'Spatial Proteomics',\n",
       " 3: 'Spatial Transcriptomics',\n",
       " 4: 'Spatial Transcriptomics',\n",
       " 5: 'Spatial Transcriptomics',\n",
       " 6: 'Spatial Transcriptomics',\n",
       " 8: 'Spatial Transcriptomics',\n",
       " 9: 'Spatial Transcriptomics',\n",
       " 10: 'Spatial Transcriptomics',\n",
       " 11: 'Spatial Proteomics',\n",
       " 12: 'Spatial Proteomics',\n",
       " 13: 'Spatial Proteomics',\n",
       " 14: 'Spatial Proteomics',\n",
       " 15: 'Spatial Proteomics',\n",
       " 16: 'Spatial Metabolomics',\n",
       " 17: 'Spatial Metabolomics',\n",
       " 18: 'Spatial Metabolomics',\n",
       " 20: 'Spatial Genomics',\n",
       " 22: 'Spatial MultiOmics',\n",
       " 24: 'Spatial MultiOmics',\n",
       " 25: 'Spatial Transcriptomics',\n",
       " 26: 'Spatial Transcriptomics',\n",
       " 27: 'Spatial Transcriptomics',\n",
       " 28: 'Spatial Transcriptomics',\n",
       " 29: 'Spatial Transcriptomics'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biotech_type_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.Biotech_ID[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_biotech_type = []\n",
    "for i in range(dataset.shape[0]):\n",
    "    dataset_biotech_type.append(biotech_type_dict[dataset.Biotech_ID[i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset['biotech_type']=dataset_biotech_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.insert(loc=5,column='biotech_type',value=dataset_biotech_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name_Short</th>\n",
       "      <th>Name_Long</th>\n",
       "      <th>Biotech_ID</th>\n",
       "      <th>biotech_type</th>\n",
       "      <th>doi</th>\n",
       "      <th>Date</th>\n",
       "      <th>Country</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Author</th>\n",
       "      <th>Access</th>\n",
       "      <th>Zip_Path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chen2021dissecting</td>\n",
       "      <td>Dissecting mammalian spermatogenesis using spa...</td>\n",
       "      <td>25</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>https://doi.org/10.1016/j.celrep.2021.109915</td>\n",
       "      <td>2021</td>\n",
       "      <td>USA</td>\n",
       "      <td>Cell Reports</td>\n",
       "      <td>Chen, H.\\nMurray, E.\\nSinha, A.\\nLaumas, A.\\nL...</td>\n",
       "      <td>https://www.dropbox.com/s/ygzpj0d0oh67br0/Test...</td>\n",
       "      <td>demo_data/zip/chen2021dissecting.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>stickels2020highly</td>\n",
       "      <td>Highly sensitive spatial transcriptomics at ne...</td>\n",
       "      <td>25</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1038/s41587-020-0739-1</td>\n",
       "      <td>2020</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Biotechnology</td>\n",
       "      <td>Stickels, Robert R.\\nMurray, Evan\\nKumar, Pawa...</td>\n",
       "      <td>https://singlecell.broadinstitute.org/single_c...</td>\n",
       "      <td>demo_data/zip/maynard2021transcriptome.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>rodriques2019slide</td>\n",
       "      <td>Slide-seq: A scalable technology for measuring...</td>\n",
       "      <td>1</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1126/science.aaw1219</td>\n",
       "      <td>2019</td>\n",
       "      <td>USA</td>\n",
       "      <td>Science</td>\n",
       "      <td>Rodriques, S. G.\\nStickels, R. R.\\nGoeva, A.\\n...</td>\n",
       "      <td>https://portals.broadinstitute.org/single_ cel...</td>\n",
       "      <td>demo_data/zip/hunter2021spatially.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>parigi2022the</td>\n",
       "      <td>The spatial transcriptomic landscape of the he...</td>\n",
       "      <td>0</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1038/s41467-022-28497-0</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Communications</td>\n",
       "      <td>Parigi, Sara M.\\nLarsson, Ludvig\\nDas, Srustid...</td>\n",
       "      <td>GSE190595</td>\n",
       "      <td>demo_data/zip/keren2018structured.zip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>guilliams2022spatial</td>\n",
       "      <td>Spatial proteogenomics reveals distinct and ev...</td>\n",
       "      <td>0</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1016/j.cell.2021.12.018</td>\n",
       "      <td>2022</td>\n",
       "      <td>Belgium</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Guilliams, M.\\nBonnardel, J.\\nHaest, B.\\nVande...</td>\n",
       "      <td>GSE192742</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>84</td>\n",
       "      <td>117</td>\n",
       "      <td>Marshall2022High_mouse</td>\n",
       "      <td>High-resolution Slide-seqV2 spatial transcript...</td>\n",
       "      <td>25</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>https://doi.org/10.1016/j.isci.2022.104097</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>iScience</td>\n",
       "      <td>Marshall, Jamie L\\nNoel, Teia\\nWang, Qingbo S\\...</td>\n",
       "      <td>cellxgene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>85</td>\n",
       "      <td>118</td>\n",
       "      <td>Barkley2022Cancer</td>\n",
       "      <td>Cancer cell states recur across tumor types an...</td>\n",
       "      <td>0</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1038/s41588-022-01141-9</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Nature Genetics</td>\n",
       "      <td>Barkley, D.\\nMoncada, R.\\nPour, M.\\nLiberman, ...</td>\n",
       "      <td>GSE203612</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>86</td>\n",
       "      <td>119</td>\n",
       "      <td>Biermann2022Dissecting</td>\n",
       "      <td>Dissecting the treatment-naive ecosystem of hu...</td>\n",
       "      <td>25</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1016/j.cell.2022.06.007</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Biermann, J.\\nMelms, J. C.\\nAmin, A. D.\\nWang,...</td>\n",
       "      <td>GSE185386</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>87</td>\n",
       "      <td>120</td>\n",
       "      <td>Wang2018Three_1k</td>\n",
       "      <td>Three-dimensional intact-tissue sequencing of ...</td>\n",
       "      <td>9</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "      <td>10.1126/science.aat5691</td>\n",
       "      <td>2018</td>\n",
       "      <td>USA</td>\n",
       "      <td>Science</td>\n",
       "      <td>Wang, X.\\nAllen, W. E.\\nWright, M. A.\\nSylwest...</td>\n",
       "      <td>http://starmapresources.com/data</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>88</td>\n",
       "      <td>121</td>\n",
       "      <td>Beechem2022High</td>\n",
       "      <td>High-plex Multiomic Analysis in FFPE at Subcel...</td>\n",
       "      <td>24</td>\n",
       "      <td>Spatial MultiOmics</td>\n",
       "      <td>https://doi.org/10.1101/2021.11.03.467020</td>\n",
       "      <td>2022</td>\n",
       "      <td>USA</td>\n",
       "      <td>bioRxiv</td>\n",
       "      <td>Beechem, Joseph M</td>\n",
       "      <td>https://nanostring.com/products/cosmx-spatial-...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0   ID              Name_Short  \\\n",
       "0            0    0      chen2021dissecting   \n",
       "1            1    1      stickels2020highly   \n",
       "2            2    2      rodriques2019slide   \n",
       "3            3    3           parigi2022the   \n",
       "4            4    4    guilliams2022spatial   \n",
       "..         ...  ...                     ...   \n",
       "84          84  117  Marshall2022High_mouse   \n",
       "85          85  118       Barkley2022Cancer   \n",
       "86          86  119  Biermann2022Dissecting   \n",
       "87          87  120        Wang2018Three_1k   \n",
       "88          88  121         Beechem2022High   \n",
       "\n",
       "                                            Name_Long  Biotech_ID  \\\n",
       "0   Dissecting mammalian spermatogenesis using spa...          25   \n",
       "1   Highly sensitive spatial transcriptomics at ne...          25   \n",
       "2   Slide-seq: A scalable technology for measuring...           1   \n",
       "3   The spatial transcriptomic landscape of the he...           0   \n",
       "4   Spatial proteogenomics reveals distinct and ev...           0   \n",
       "..                                                ...         ...   \n",
       "84  High-resolution Slide-seqV2 spatial transcript...          25   \n",
       "85  Cancer cell states recur across tumor types an...           0   \n",
       "86  Dissecting the treatment-naive ecosystem of hu...          25   \n",
       "87  Three-dimensional intact-tissue sequencing of ...           9   \n",
       "88  High-plex Multiomic Analysis in FFPE at Subcel...          24   \n",
       "\n",
       "               biotech_type                                           doi  \\\n",
       "0   Spatial Transcriptomics  https://doi.org/10.1016/j.celrep.2021.109915   \n",
       "1   Spatial Transcriptomics                     10.1038/s41587-020-0739-1   \n",
       "2   Spatial Transcriptomics                       10.1126/science.aaw1219   \n",
       "3   Spatial Transcriptomics                    10.1038/s41467-022-28497-0   \n",
       "4   Spatial Transcriptomics                    10.1016/j.cell.2021.12.018   \n",
       "..                      ...                                           ...   \n",
       "84  Spatial Transcriptomics    https://doi.org/10.1016/j.isci.2022.104097   \n",
       "85  Spatial Transcriptomics                    10.1038/s41588-022-01141-9   \n",
       "86  Spatial Transcriptomics                    10.1016/j.cell.2022.06.007   \n",
       "87  Spatial Transcriptomics                       10.1126/science.aat5691   \n",
       "88       Spatial MultiOmics     https://doi.org/10.1101/2021.11.03.467020   \n",
       "\n",
       "    Date  Country                Journal  \\\n",
       "0   2021      USA           Cell Reports   \n",
       "1   2020      USA   Nature Biotechnology   \n",
       "2   2019      USA                Science   \n",
       "3   2022      USA  Nature Communications   \n",
       "4   2022  Belgium                   Cell   \n",
       "..   ...      ...                    ...   \n",
       "84  2022      USA               iScience   \n",
       "85  2022      USA        Nature Genetics   \n",
       "86  2022      USA                   Cell   \n",
       "87  2018      USA                Science   \n",
       "88  2022      USA                bioRxiv   \n",
       "\n",
       "                                               Author  \\\n",
       "0   Chen, H.\\nMurray, E.\\nSinha, A.\\nLaumas, A.\\nL...   \n",
       "1   Stickels, Robert R.\\nMurray, Evan\\nKumar, Pawa...   \n",
       "2   Rodriques, S. G.\\nStickels, R. R.\\nGoeva, A.\\n...   \n",
       "3   Parigi, Sara M.\\nLarsson, Ludvig\\nDas, Srustid...   \n",
       "4   Guilliams, M.\\nBonnardel, J.\\nHaest, B.\\nVande...   \n",
       "..                                                ...   \n",
       "84  Marshall, Jamie L\\nNoel, Teia\\nWang, Qingbo S\\...   \n",
       "85  Barkley, D.\\nMoncada, R.\\nPour, M.\\nLiberman, ...   \n",
       "86  Biermann, J.\\nMelms, J. C.\\nAmin, A. D.\\nWang,...   \n",
       "87  Wang, X.\\nAllen, W. E.\\nWright, M. A.\\nSylwest...   \n",
       "88                                  Beechem, Joseph M   \n",
       "\n",
       "                                               Access  \\\n",
       "0   https://www.dropbox.com/s/ygzpj0d0oh67br0/Test...   \n",
       "1   https://singlecell.broadinstitute.org/single_c...   \n",
       "2   https://portals.broadinstitute.org/single_ cel...   \n",
       "3                                           GSE190595   \n",
       "4                                           GSE192742   \n",
       "..                                                ...   \n",
       "84                                          cellxgene   \n",
       "85                                          GSE203612   \n",
       "86                                          GSE185386   \n",
       "87                   http://starmapresources.com/data   \n",
       "88  https://nanostring.com/products/cosmx-spatial-...   \n",
       "\n",
       "                                      Zip_Path  \n",
       "0         demo_data/zip/chen2021dissecting.zip  \n",
       "1   demo_data/zip/maynard2021transcriptome.zip  \n",
       "2        demo_data/zip/hunter2021spatially.zip  \n",
       "3        demo_data/zip/keren2018structured.zip  \n",
       "4                                          NaN  \n",
       "..                                         ...  \n",
       "84                                         NaN  \n",
       "85                                         NaN  \n",
       "86                                         NaN  \n",
       "87                                         NaN  \n",
       "88                                         NaN  \n",
       "\n",
       "[89 rows x 13 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_trans = dataset[dataset.biotech_type=='Spatial Transcriptomics']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_prot = dataset[dataset.biotech_type=='Spatial Proteomics']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Belgium', 'Canada', 'China', 'France', 'Japan', 'Spain', 'Sweden',\n",
       "        'Switzerland', 'UK', 'USA'], dtype=object),\n",
       " array([ 1,  1,  6,  1,  1,  1, 15,  1,  5, 37]))"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dataset_trans['Country'],return_counts=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Switzerland', 'USA'], dtype=object), array([2, 8]))"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dataset_prot['Country'],return_counts=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Belgium:1\n",
      "Canada:1\n",
      "China:6\n",
      "France:1\n",
      "Japan:1\n",
      "Spain:1\n",
      "Sweden:15\n",
      "Switzerland:1\n",
      "UK:5\n",
      "USA:37\n"
     ]
    }
   ],
   "source": [
    "biotech_name_list,biotech_name_vals=np.unique(dataset_trans['Country'],return_counts=1)\n",
    "for i in range(len(biotech_name_vals)):\n",
    "    print(f'{biotech_name_list[i]}:{biotech_name_vals[i]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Switzerland:2\n",
      "USA:8\n"
     ]
    }
   ],
   "source": [
    "biotech_name_list,biotech_name_vals = np.unique(dataset_prot['Country'],return_counts=1)\n",
    "for i in range(len(biotech_name_vals)):\n",
    "    print(f'{biotech_name_list[i]}:{biotech_name_vals[i]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Belgium', 'China', 'Sweden', 'UK', 'USA'], dtype=object),\n",
       " array([ 1,  4, 13,  3, 21]))"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dataset_trans['Country'],return_counts=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Switzerland', 'USA'], dtype=object), array([ 2, 11]))"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(dataset_prot['Country'],return_counts=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SODB",
   "language": "python",
   "name": "sodb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
