{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualization and analysis of gene expression in tissue sections by spatial transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ShortName: he2020integrating"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Steps of processing the data from raw to Anndata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1, Download the raw data from https://data.mendeley.com/datasets/29ntw7sh4r/5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data path of the ST data\n",
    "data_path = './data/BreastCancer2021'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read counts\n",
    "def read_count(count_file):\n",
    "    print('reading ' + count_file)\n",
    "    count = pd.read_csv(f'{data_path}/{count_file}', sep='\\t')\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read spot_coordinates\n",
    "def read_spot_coordinates(spot_coordinates_file):\n",
    "    print('reading ' + spot_coordinates_file)\n",
    "    spot_coordinates = pd.read_csv(f'{data_path}/{spot_coordinates_file}')\n",
    "    spot_coordinates = spot_coordinates.rename({'Unnamed: 0': 'coord_ID'}, axis='columns')\n",
    "    spot_coordinates = spot_coordinates.set_index('coord_ID')\n",
    "    return spot_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# read tumor_annotation\n",
    "def read_tumor_annotation(tumor_annotation_file):\n",
    "    print('reading ' + tumor_annotation_file)\n",
    "    tumor_annotation = pd.read_csv(f'{data_path}/{tumor_annotation_file}', sep='\\t')\n",
    "    tumor_annotation['cell_ID'] = tumor_annotation['xcoord']\n",
    "    tumor_annotation['xcoord'] = tumor_annotation['ycoord']\n",
    "    tumor_annotation['ycoord'] = tumor_annotation['lab']\n",
    "    tumor_annotation['lab'] = tumor_annotation['tumor']\n",
    "    tumor_annotation['tumor'] = tumor_annotation['Unnamed: 4']\n",
    "    tumor_annotation = tumor_annotation.drop(['Unnamed: 4'], axis=1)\n",
    "    tumor_annotation['coord_ID'] = tumor_annotation.apply(lambda x: str(x['xcoord']) + 'x' + str(x['ycoord']), axis=1)\n",
    "    tumor_annotation = tumor_annotation.set_index('coord_ID')\n",
    "    return tumor_annotation\n",
    "# Unnamed: 0 x y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>patient</th>\n",
       "      <th>replicate</th>\n",
       "      <th>count_matrix</th>\n",
       "      <th>histology_image</th>\n",
       "      <th>spot_coordinates</th>\n",
       "      <th>tumor_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HER2_luminal</td>\n",
       "      <td>BC23287</td>\n",
       "      <td>C1</td>\n",
       "      <td>BC23287_C1_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23287_C1.jpg</td>\n",
       "      <td>spots_BT23287_C1.csv.gz</td>\n",
       "      <td>BC23287_C1_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HER2_luminal</td>\n",
       "      <td>BC23287</td>\n",
       "      <td>C2</td>\n",
       "      <td>BC23287_C2_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23287_C2.jpg</td>\n",
       "      <td>spots_BT23287_C2.csv.gz</td>\n",
       "      <td>BC23287_C2_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HER2_luminal</td>\n",
       "      <td>BC23287</td>\n",
       "      <td>D1</td>\n",
       "      <td>BC23287_D1_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23287_D1.jpg</td>\n",
       "      <td>spots_BT23287_D1.csv.gz</td>\n",
       "      <td>BC23287_D1_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HER2_luminal</td>\n",
       "      <td>BC23450</td>\n",
       "      <td>D2</td>\n",
       "      <td>BC23450_D2_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23450_D2.jpg</td>\n",
       "      <td>spots_BT23450_D2.csv.gz</td>\n",
       "      <td>BC23450_D2_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HER2_luminal</td>\n",
       "      <td>BC23450</td>\n",
       "      <td>E1</td>\n",
       "      <td>BC23450_E1_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23450_E1.jpg</td>\n",
       "      <td>spots_BT23450_E1.csv.gz</td>\n",
       "      <td>BC23450_E1_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TNBC</td>\n",
       "      <td>BC23377</td>\n",
       "      <td>C2</td>\n",
       "      <td>BC23377_C2_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23377_C2.jpg</td>\n",
       "      <td>spots_BT23377_C2.csv.gz</td>\n",
       "      <td>BC23377_C2_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TNBC</td>\n",
       "      <td>BC23377</td>\n",
       "      <td>D1</td>\n",
       "      <td>BC23377_D1_stdata.tsv.gz</td>\n",
       "      <td>HE_BT23377_D1.jpg</td>\n",
       "      <td>spots_BT23377_D1.csv.gz</td>\n",
       "      <td>BC23377_D1_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>TNBC</td>\n",
       "      <td>BC23803</td>\n",
       "      <td>D2</td>\n",
       "      <td>BC23803_D2_stdata.tsv.gz</td>\n",
       "      <td>HE_BC23803_D2.jpg</td>\n",
       "      <td>spots_BC23803_D2.csv.gz</td>\n",
       "      <td>BC23803_D2_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>TNBC</td>\n",
       "      <td>BC23803</td>\n",
       "      <td>E1</td>\n",
       "      <td>BC23803_E1_stdata.tsv.gz</td>\n",
       "      <td>HE_BC23803_E1.jpg</td>\n",
       "      <td>spots_BC23803_E1.csv.gz</td>\n",
       "      <td>BC23803_E1_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>TNBC</td>\n",
       "      <td>BC23803</td>\n",
       "      <td>E2</td>\n",
       "      <td>BC23803_E2_stdata.tsv.gz</td>\n",
       "      <td>HE_BC23803_E2.jpg</td>\n",
       "      <td>spots_BC23803_E2.csv.gz</td>\n",
       "      <td>BC23803_E2_Coords.tsv.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            type  patient replicate              count_matrix  \\\n",
       "0   HER2_luminal  BC23287        C1  BC23287_C1_stdata.tsv.gz   \n",
       "1   HER2_luminal  BC23287        C2  BC23287_C2_stdata.tsv.gz   \n",
       "2   HER2_luminal  BC23287        D1  BC23287_D1_stdata.tsv.gz   \n",
       "3   HER2_luminal  BC23450        D2  BC23450_D2_stdata.tsv.gz   \n",
       "4   HER2_luminal  BC23450        E1  BC23450_E1_stdata.tsv.gz   \n",
       "..           ...      ...       ...                       ...   \n",
       "63          TNBC  BC23377        C2  BC23377_C2_stdata.tsv.gz   \n",
       "64          TNBC  BC23377        D1  BC23377_D1_stdata.tsv.gz   \n",
       "65          TNBC  BC23803        D2  BC23803_D2_stdata.tsv.gz   \n",
       "66          TNBC  BC23803        E1  BC23803_E1_stdata.tsv.gz   \n",
       "67          TNBC  BC23803        E2  BC23803_E2_stdata.tsv.gz   \n",
       "\n",
       "      histology_image         spot_coordinates          tumor_annotation  \n",
       "0   HE_BT23287_C1.jpg  spots_BT23287_C1.csv.gz  BC23287_C1_Coords.tsv.gz  \n",
       "1   HE_BT23287_C2.jpg  spots_BT23287_C2.csv.gz  BC23287_C2_Coords.tsv.gz  \n",
       "2   HE_BT23287_D1.jpg  spots_BT23287_D1.csv.gz  BC23287_D1_Coords.tsv.gz  \n",
       "3   HE_BT23450_D2.jpg  spots_BT23450_D2.csv.gz  BC23450_D2_Coords.tsv.gz  \n",
       "4   HE_BT23450_E1.jpg  spots_BT23450_E1.csv.gz  BC23450_E1_Coords.tsv.gz  \n",
       "..                ...                      ...                       ...  \n",
       "63  HE_BT23377_C2.jpg  spots_BT23377_C2.csv.gz  BC23377_C2_Coords.tsv.gz  \n",
       "64  HE_BT23377_D1.jpg  spots_BT23377_D1.csv.gz  BC23377_D1_Coords.tsv.gz  \n",
       "65  HE_BC23803_D2.jpg  spots_BC23803_D2.csv.gz  BC23803_D2_Coords.tsv.gz  \n",
       "66  HE_BC23803_E1.jpg  spots_BC23803_E1.csv.gz  BC23803_E1_Coords.tsv.gz  \n",
       "67  HE_BC23803_E2.jpg  spots_BC23803_E2.csv.gz  BC23803_E2_Coords.tsv.gz  \n",
       "\n",
       "[68 rows x 7 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read meta data\n",
    "meta_path = 'metadata.csv'\n",
    "metadata = pd.read_csv(f'{data_path}/{meta_path}')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, matadata_row in metadata.iterrows():\n",
    "    count_file = matadata_row['count_matrix']\n",
    "    spot_coordinates_file =  matadata_row['spot_coordinates']\n",
    "    tumor_annotation_file = matadata_row['tumor_annotation']\n",
    "    type_patient = matadata_row['type']\n",
    "\n",
    "\n",
    "    #preprocess the data\n",
    "    count = read_count(count_file)\n",
    "    spot_coordinates = read_spot_coordinates(spot_coordinates_file)\n",
    "    tumor_annotation = read_tumor_annotation(tumor_annotation_file)\n",
    "\n",
    "    # get geneID\n",
    "    gene_id_array = count.columns\n",
    "    # remove \"Unnamed: 0\"\n",
    "    gene_id_array = gene_id_array[1:]\n",
    "\n",
    "    # get cellID\n",
    "    count = count.rename({'Unnamed: 0': 'coord_ID'}, axis='columns')\n",
    "    cell_id_array = count['coord_ID']\n",
    "\n",
    "    # get count_matrix\n",
    "    count_X = count.values[:,1:].astype('float')\n",
    "\n",
    "    # adata\n",
    "    adata = ad.AnnData(count_X)\n",
    "    adata.var_names = gene_id_array\n",
    "    adata.obs_names = cell_id_array\n",
    "\n",
    "    # merge tumor_annotation\n",
    "    adata.obs = adata.obs.merge(tumor_annotation[['lab', 'cell_ID', 'tumor']], how='left', on='coord_ID')\n",
    "    # merge spot_coordinates\n",
    "    adata.obs = adata.obs.merge(spot_coordinates, how='left', on='coord_ID')\n",
    "    adata.obs = adata.obs.rename({'X': 'xcoord'}, axis='columns')\n",
    "    adata.obs = adata.obs.rename({'Y': 'ycoord'}, axis='columns')\n",
    "\n",
    "    # get spot coordinate\n",
    "    count['x'] = count['coord_ID'].apply(lambda x: x.split('x')[0])\n",
    "    count['y'] = count['coord_ID'].apply(lambda x: x.split('x')[1])\n",
    "    count = count.set_index('coord_ID')\n",
    "    adata.obsm['spatial'] = count.loc[adata.obs_names].loc[:,['x','y']]\n",
    "\n",
    "    write_file_name = 'staahl2016visualization' + '_' + type_patient + '_' + '_'.join(count_file.split('_')[:-1]) + '.h5ad'\n",
    "    adata.write_h5ad(f'{data_path}/h5ad/{write_file_name}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "98b0a9b7b4eaaa670588a142fd0a9b87eaafe866f1db4228be72b4211d12040f"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
